### PR TITLE
feat: theme selector

### LINKED ISSUE #598: Make dark/light mode configurable
Currently we use the system's dark/light settings. Can this also be a custom Alby setting? (mode: light/dark/system?) 

https://feedback.getalby.com/feature-requests-/posts/i-would-like-no-dark-mode-option-on-alby

### PR DESCRIPTION
**Pull request by secondl1ght**
#### Link this PR to an issue
Fixes #598 
Closes #618 

#### Type of change (Remove other not matching type)

- New feature (non-breaking change which adds functionality)

#### Describe the changes you have made in this PR -
Added a theme selector to the Settings screen which allows the user to choose dark, light or system theme for Alby. This settings is stored and used on all screens for Alby, I also added a transition animation for when you change the selection on the Settings page.

#### Screenshots of the changes (If any) -
image

#### How Has This Been Tested?
Tested all the 3 different theme options on all the screens within Alby. Also ran `console.log` tests to make sure the Settings were persistently updated.

#### Checklist:

- [x] My code follows the style guidelines of this project and performed a self-review of my own code
- [x] New and existing tests pass locally with my changes
- [x] I checked if I need to make corresponding changes to the documentation (and made those changes if needed)

### CODE REVIEW

File: src/app/context/AuthContext.tsx [hunk 1]
```diff
@@ -27,6 +27,24 @@ interface AuthContextType {
 const AuthContext = createContext({} as AuthContextType);
 
 export function AuthProvider({ children }: { children: React.ReactNode }) {
+  //theme
+  useEffect(() => {
+    api.getSettings().then((response) => {
```
Comment (escapedcat) [inline src/app/context/AuthContext.tsx:32]: More like a general question: Does Alby prefer `.then()` or `async/await` or just doesn't matter?

File: src/app/context/AuthContext.tsx [hunk 2]
```diff
@@ -27,6 +27,24 @@ interface AuthContextType {
...
+      const settings = response;
```
Comment (escapedcat) [inline src/app/context/AuthContext.tsx:33]: Cab it be worth typing this?

File: src/app/context/AuthContext.tsx (hunk 1 above)
Comment (escapedcat) [inline src/app/context/AuthContext.tsx:32]: Does it make sense to move this into a i.e. util function because it looks like the same code is being used in 3 places?

File: src/app/screens/Settings.tsx [hunk 3]
```diff
@@ -140,6 +157,26 @@ function Settings() {
...
+              <Select
+                name="theme"
+                value={settings.theme}
+                onChange={(ev) => {
+                  saveSetting({
+                    theme: ev.target.value,
+                  });
+                }}
+              >
+                <option value="dark">Dark</option>
```
Comment (escapedcat) [inline src/app/screens/Settings.tsx:173]: Now that we have translaitons in should each ne PR already be forced to use translation-placeholders instead of strings?

File: src/app/context/AuthContext.tsx (hunk 1 above)
Comment (secondl1ght) [reply src/app/context/AuthContext.tsx:32]: `.then()` was already being used for this function previously so that's why I used it, but it looks like both then and `await` are used in the codebase. Not sure which one is preferred though.

File: src/app/context/AuthContext.tsx (hunk 2 above)
Comment (secondl1ght) [reply src/app/context/AuthContext.tsx:33]: You mean typing the response of the `getSettings` call?

File: src/app/context/AuthContext.tsx (hunk 1 above)
Comment (secondl1ght) [reply src/app/context/AuthContext.tsx:32]: I thought about what I could do to make this more DRY, but didn't know you could turn a `useEffect` hook into a function. It is possible?

File: src/app/screens/Settings.tsx (hunk 3 above)
Comment (secondl1ght) [reply src/app/screens/Settings.tsx:173]: Ok I can update this with the translations.

Comment (secondl1ght) [reply src/app/screens/Settings.tsx:173]: I just noticed none of the other items on this page have the translations yet it looks like, should we put in a separate PR to update the entire app with translations?

File: src/app/context/AuthContext.tsx (hunk 1 above)
Comment (escapedcat) [reply src/app/context/AuthContext.tsx:32]: Yeah, just noticed this. Might be good to think about a preferred way.  
@dylancom what do you think?

Definitely not blocking this.

Comment (escapedcat) [reply src/app/context/AuthContext.tsx:32]: Not sure, I assumed you can use a function in that place.

File: src/app/context/AuthContext.tsx (hunk 2 above)
Comment (escapedcat) [reply src/app/context/AuthContext.tsx:33]: Yes, would that work?

File: src/app/screens/Settings.tsx (hunk 3 above)
Comment (escapedcat) [reply src/app/screens/Settings.tsx:173]: Created #781 for that.  
But it might be good to just start with this and do it step by step already.

File: src/app/context/AuthContext.tsx (hunk 2 above)
Comment (secondl1ght) [reply src/app/context/AuthContext.tsx:33]: I think it would work but the variables that get sent from the API call are already typed so it might be redundant?

File: src/app/context/AuthContext.tsx (hunk 1 above)
Comment (secondl1ght) [reply src/app/context/AuthContext.tsx:32]: Ok I have never done that before but i'll give it a try and see if it works.

Comment (secondl1ght) [reply src/app/context/AuthContext.tsx]: Looks like I was able to get it working, see latest commit.

File: src/app/utils/index.ts [hunk 4]
```diff
@@ -1,3 +1,25 @@
...
+
 export function classNames(...classes: (string | boolean)[]) {
   return classes.filter(Boolean).join(" ");
 }
+
+export function useThemeEffect(dependencies: undefined | string) {
+  useEffect(() => {
+    api.getSettings().then((response) => {
+      const settings = response;
+      if (settings.theme === "system") {
```
Comment (bumi) [inline src/app/utils/index.ts:12]: can you add a comment here what this is doing?

File: src/app/router/Welcome/Welcome.tsx [hunk 5]
```diff
@@ -53,6 +53,12 @@ function App() {
   const { t } = useTranslation(["welcome"]);
   const location = useLocation();
   const routesElement = useRoutes(routes);
+  //theme
+  useEffect(() => {
+    if (window.matchMedia("(prefers-color-scheme: dark)").matches) {
```
Comment (bumi) [inline src/app/router/Welcome/Welcome.tsx:58]: why is this here but not on the other router/* ?

**Review by bumi — CHANGES_REQUESTED** [review]
Where do those `transition-colors` need to be added?

File: src/app/utils/index.ts (hunk 4 above)
Comment (secondl1ght) [reply src/app/utils/index.ts:14]: Ok I added some comments to explain how the function works, please let me know if this is what you wanted.

File: src/app/router/Welcome/Welcome.tsx (hunk 5 above)
Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx:58]: This screen needed its own `useEffect` function because I could not use the one I wrote in `utils` for the other screens. The reason is that for the other screens I am checking the `Settings` for the theme selection. But on the Welcome screen the settings have not yet been initiated by the initial user on boarding. So in this case we only need to check what the users system theme settings are and not use the API call that doesn't exist yet. After the user sets up their first wallet they then have access to the Settings screen and can adjust the theme from there, which is why all of the screens after the initial on boarding check the Settings API.

Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx:58]: Does this make sense?

**Comment by secondl1ght** [comment]
> Where do those `transition-colors` need to be added?

I added this so that there was a smoother transition effect when the user switches between dark/light themes. It is not necessary though, I can remove it if you want?

File: src/app/router/Welcome/Welcome.tsx (hunk 5 above)
Comment (bumi) [reply src/app/router/Welcome/Welcome.tsx:58]: hmm. yes makes sense. 
but does it hurt doing the settings call? When there is no special setting then it just uses the default system theme, doesn't it? 

having less code is always better - less things that break :)

Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx:58]: I think I tired using the same function for this screen but it wasn't working but I will t

> > I added this so that there was a smoother transition effect when the user switches between dark/light themes. It is not necessary though, I can remove it if you want?
> 
> I don't know. I just fear that this will be forgotten when developing new screens.

Yeah it's not a big deal, most people wouldn't notice it to be honest. I will remove.

**Comment by secondl1ght** [comment]
> > I added this so that there was a smoother transition effect when the user switches between dark/light themes. It is not necessary though, I can remove it if you want?
> 
> I don't know. I just fear that this will be forgotten when developing new screens.

Yeah it's not a big deal, most people wouldn't notice it to be honest. I will remove.

File: src/app/router/Welcome/Welcome.tsx (hunk 5 above)
Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx:58]: I think I tried this originally but it was not working - I will try again.

Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx:58]: So yes you are right about defaulting to system theme but for some reason when I first launch Alby this code does not work. There is an error in the devtools console, but when I refresh the page then it starts working. So I think in production it may also not work when you first launch Alby. Maybe it is because you cant use the API call right when Alby opens for the first time? Maybe @dylancom would know more but personally I think having one unique `useEffect` call here to get it working is not that much of an issue.

This is the error it gives:
`Uncaught (in promise) Error: Could not establish connection. Receiving end does not exist.
    at wrappedSendMessageCallback (browser-polyfill.js:1187:1)`

Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx:58]: I think it means that `api.getSettings` does not exist yet when you first launch the extension.

Comment (bumi) [reply src/app/router/Welcome/Welcome.tsx:58]: maybe we first need to debug and understand that error before drawing conclusions.

Comment (secondl1ght) [reply src/app/router/Welcome/Welcome.tsx]: Ok I did some more debugging and was able to add a new parameter to the function to check if it is on the initial Welcome page. If it is then it will run the code without the API call, all the other pages will use the API call like before.

Comment (bumi) [reply src/app/router/Welcome/Welcome.tsx]: I really think we should first understand that error and then see how we can fix it instead of hacking around it. 
why is doing that call to load the settings a problem? I don't know the setup on this right now but I am wondering why the different pages/screens need to be treated differently. 

though it is great that it works, magic parameters make the code harder to read. people who don't have the background or in a few months don't know what the `true/false` means when reading `useThemeEffect(noDependencies, false);`

File: src/app/utils/index.ts [hunk 6]
```diff
@@ -1,3 +1,32 @@
+import { useEffect } from "react";
+import api from "../../common/lib/api";
+
 export function classNames(...classes: (string | boolean)[]) {
   return classes.filter(Boolean).join(" ");
 }
+
+export function useThemeEffect(dependencies: undefined | string) {
```
Comment (dylancom) [inline src/app/utils/index.ts:8]: We should move this into a "hooks" folder.

**Review by dylancom — CHANGES_REQUESTED** [review]

File: src/app/context/AuthContext.tsx (hunk 1 above)
Comment (dylancom) [reply src/app/context/AuthContext.tsx]: I think in general async/await is preferred in terms or readability and to avoid callback hell.
It's currently a bit of a mix of both in the codebase.

**Comment by dylancom** [comment]
Seems to work fine on my end (no error)

File: src/app/router/Welcome/Welcome.tsx (hunk 5 above)
Comment (dylancom) [reply src/app/router/Welcome/Welcome.tsx]: The error was caused by onMessage listeners being registered after opening welcome.html. So I made sure that welcome is now opened after initialization has finished.

**Comment by secondl1ght** [comment]
Ok thanks a lot to @dylancom for refactoring this and figuring out that hard to solve bug!! 🙌🏼 

I believe this PR is ready to be merged now.
